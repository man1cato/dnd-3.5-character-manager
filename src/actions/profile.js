import database from '../firebase/firebase'


//SET PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
});

//READ PROFILE FROM FIREBASE
export const startSetProfile = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val();
            dispatch(setProfile(id, profile));
        });
    };
};

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
});

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates));
        });
    };
};

//DELETE PROFILE FROM STORE
export const deleteProfile = () => ({
    type: 'DELETE_PROFILE'
})

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
