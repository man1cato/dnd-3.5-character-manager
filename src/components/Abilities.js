import React from 'react';

const Abilities = ({ abilities, handleChange }) => (
	<div className="grid grid--abilities">
		<h5 className="grid__col1">Ability</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col1" key={i}>{abilities[ability].name}</div>
		)}

		<h5 className="grid__col2">Score</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col2" key={i}>{abilities[ability].score}</div>
		)}

		<h5 className="grid__col3">Mod</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col3" key={i}>{abilities[ability].mod}</div>
		)}

		<h5 className="grid__col4">Temp Score</h5>
		{Object.keys(abilities).map((ability, i) =>
			<input
			className="grid__col4"
			key={i}
			id={ability}
			name="abilities"
			type="text"
			value={abilities[ability].tempScore}
			onChange={handleChange}
			/>
		)}

		<h5 className="grid__col5">Temp Mod</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col5" key={i}>{abilities[ability].tempMod}</div>
		)}
	</div>
)

export default Abilities;
