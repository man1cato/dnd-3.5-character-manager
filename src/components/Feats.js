import React from 'react'
import {connect} from 'react-redux'
import _ from 'lodash'
import FeatModal from './FeatModal'

export class Feats extends React.Component{
	state = {
		selected: undefined
	}  

	handleOpenModal = (e) => {
		const featId = e.target.id
		const selected = this.props.feats[featId]
		this.setState({selected})
	}

	handleCloseModal = () => {
		this.setState({selected: undefined})
	}
	
	render () {    
		const feats = _.orderBy(this.props.featIds.map((featId) => ({
			...this.props.feats[featId],
			id: featId
		})), ['name'], ['asc'])

		return (
			<div>
				{feats.map((feat, i) => (
					<button
						className="button--link"
						id={feat.id}
						key={`feat${i}`}
						onClick={this.handleOpenModal}
					>
						{feat.name}
					</button>
				))}

				<FeatModal 
					selected={this.state.selected}
					handleCloseModal={this.handleCloseModal}
				/>
			</div>
		)
	}
}


const mapStateToProps = (state) => ({
  	feats: state.feats
})

export default connect(mapStateToProps)(Feats)