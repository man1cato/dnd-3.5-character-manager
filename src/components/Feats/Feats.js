import React, { useState } from 'react'
import { connect } from 'react-redux'
import _ from 'lodash'

import FeatModal from '../FeatModal'


export const Feats = (props) => { 
	const [selected, setSelected] = useState(undefined)

	const feats = _.sortBy(props.featIds.map((featId) => ({
		...props.feats[featId],
		id: featId
	})), ['name'])

	return (
		<div>
			{feats.map((feat, i) => (
				<button
					className="button--link"
					id={feat.id}
					key={`feat${i}`}
					type='button'
					onClick={() => setSelected(feat)}
				>
					{feat.name}
				</button>
			))}

			<FeatModal 
				selected={selected}
				handleCloseModal={() => setSelected(undefined)}
			/>
		</div>
	)
}


const mapStateToProps = (state) => ({
  	feats: state.feats
})

export default connect(mapStateToProps)(Feats)