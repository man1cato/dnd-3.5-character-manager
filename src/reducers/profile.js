export default (state = {}, action) => {
    switch (action.type) {
        case 'SET_PROFILE':
            return {
                id: action.id,
                ...action.fields
            };
        case 'EDIT_PROFILE':
            return {
                ...state,
                ...action.updates
            };
        default:
            return state;
    }
};
