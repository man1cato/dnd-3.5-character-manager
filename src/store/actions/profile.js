import database from '../../firebase/firebase'
import { startGetProfiles } from './profiles'

//SET SELECTED PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
})

//READ PROFILE FROM FIREBASE
export const startSetProfile = (uid, id) => {
    return (dispatch) => {
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val()
            dispatch(setProfile(id, profile))
        })
    }
}

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
            dispatch(startGetProfiles(uid))
            localStorage.setItem('selectedCharacterId', ref.key)
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
})

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates))
        })
    }
}

//REMOVE PROFILE FROM STORE
export const removeProfile = () => ({
    type: 'REMOVE_PROFILE'
})

//REMOVE PROFILE FROM FIREBASE
// export const startRemoveProfile = () => {
//     return (dispatch) => {
//         return database.ref(`users/${uid}/profiles/${id}`).remove().then(() => {
//             dispatch(removeProfile())
//         })
//     }
// }

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
