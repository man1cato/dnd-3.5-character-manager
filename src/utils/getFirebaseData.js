import admin from 'firebase-admin'
import serviceAccount from '../../config/dnd-3-5-character-manager-e3f58e584d80.json'

admin.initializeApp({
   credential: admin.credential.cert(serviceAccount),
   databaseURL: "https://dnd-3-5-character-manager.firebaseio.com"
 })
 
const database = admin.database()

export default async () => {
   try {
      const snapshot = await database.ref('api').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}

export const getRaces = async () => {
   try {
      const snapshot = await database.ref('api/races').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}

export const getJobClasses = async () => {
   try {
      const snapshot = await database.ref('api/jobClasses').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}

export const getSkills = async () => {
   try {
      const snapshot = await database.ref('api/skills').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}

export const getFeats = async () => {
   try {
      const snapshot = await database.ref('api/feats').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}

export const getSpecialAbilities = async () => {
   try {
      const snapshot = await database.ref('api/specialAbilities').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}

export const getItems = async () => {
   try {
      const snapshot = await database.ref('api/items').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}

export const getSpells = async () => {
   try {
      const snapshot = await database.ref('api/spells').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}