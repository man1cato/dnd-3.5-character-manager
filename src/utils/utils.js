import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])