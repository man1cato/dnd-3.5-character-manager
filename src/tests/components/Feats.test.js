import React from 'react'
import {shallow} from 'enzyme'

import { Feats } from '../../components/Feats'
import profile from '../fixtures/profile'
import { getFeats } from '../../utils/getFirebaseData'


test('should render Feats with profile data', async () => {
	const props = {
		featIds: profile.fields.feats,
		feats: await getFeats()
	}
	const wrapper = shallow(<Feats {...props} />)
	expect(wrapper).toMatchSnapshot()
})
