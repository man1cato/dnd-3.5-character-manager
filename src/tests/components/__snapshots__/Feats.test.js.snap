// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Feats with profile data 1`] = `
<div>
  <button
    className="button--link"
    id="recRjuAWe7oy6a7wY"
    key="feat0"
    onClick={[Function]}
  >
    Craft Wondrous Item
  </button>
  <button
    className="button--link"
    id="recGkxwkqppy7EYmI"
    key="feat1"
    onClick={[Function]}
  >
    Scribe Scroll
  </button>
  <button
    className="button--link"
    id="rec4KkfhKpetKIDqj"
    key="feat2"
    onClick={[Function]}
  >
    Spell Mastery
  </button>
  <FeatModal
    handleCloseModal={[Function]}
  />
</div>
`;
